2053 spin   update instrument spin phase definition

tmplummer opened the pull request:
# Change Summary

## Overview
Updates the instrument spin phase 0 to be consistent with IT desires. That is where spin-phase 0 is when the instrument boresight is closest to the ecliptic north pole.

## Updated Files
- imap_processing/spice/geometry.py
   - Add 90 degrees and take the modulus with 360 to all instrument spin phase offsets
- imap_processing/tests/spice/test_geometry.py
   - Update expected values by calculating a slightly different way than above.
- imap_processing/tests/swe/test_swe_l2.py
   - Mock the SWE offset back to the previous value since validation of spin-phase was broken by above changes.

Closes: #2053 #1023

tmplummer committed:
Add 90 degrees to each instrument spin phase offset

tmplummer committed:
Patch SWE tests back to previous spin-phase

greglucas commented on imap_processing/spice/geometry.py:
```diff
@@ -137,7 +137,9 @@ def get_spacecraft_to_instrument_spin_phase_offset(instrument: SpiceFrame) -> fl
-    to S/C CS Transformations` in document `7516-0011_drw.pdf`. These fixed
+    to S/C CS Transformations` in document `7516-0011_drw.pdf`. That Table
+    defines the angle from the spacecraft y-axis. We add 90 and take the modulous
+    with 360 in order to get the angle from the spacecraft x-axis. These fixed
     values will need to be updated based on calibration data or retrieved using
     SPICE and the latest IMAP frame kernel.
```
Do you want to implement the retrieval from the SPICE instrument kernel now that we have official alignments there?

tmplummer replied:
I went ahead and added the calculation using SPICE. @greglucas can you take another look when you get a chance?

greglucas commented on imap_processing/tests/swe/test_swe_l2.py:
```diff
@@ -256,8 +256,13 @@ def test_put_data_into_angle_bins():
 @patch("imap_data_access.processing_input.ProcessingInputCollection.get_file_paths")
+@patch(
+    "imap_processing.spice.spin.get_spacecraft_to_instrument_spin_phase_offset",
+    return_value=153 / 360,
```
This is a nice way to approach this for now. I think we might want to open an issue for @tech3371 to update the expected values at some point though...

tech3371 replied:
Sounds good. I will work with Ruth on this after launch preparation.

greglucas commented:
I also assume you know it is supposed to be +90 and not -90? I wasn't positive which direction of rotation you needed to rotate.

tech3371 reviewed (approved):
The changes looks good to me. Thank you for having workaround for SWE.

tmplummer committed:
Get instrument phase offset from SPICE
Add function for getting instrument az/el from SPICE
Add test coverage for new geometry function

tmplummer committed:
Add mock to hi test to keep spice out of it

greglucas commented on imap_processing/tests/spice/test_geometry.py:
```diff
@@ -46,27 +47,62 @@ def test_imap_state_ecliptic(imap_ena_sim_metakernel):
+    [
+        # Expected spin-phase offsets based on 7516-0011_drw.pdf
+        (SpiceFrame.IMAP_LO_BASE, (60, 0)),  # (330 + 90) % 360 = 60
+        # Note HI_45 and HI_90 appear to be swapped in imap_wkcp.tf so the
+        # expected values are swapped here.
```
This seems worth bringing up to someone...

tmplummer replied:
I am waiting for the updated imap frames kernel before I make any noise about this.

greglucas replied:
I would message Nick on the side quickly about this so it gets taken care of before we get an "updated kernel" that doesn't have this update you want!

greglucas reviewed (approved):
Looks good to me. The only downside is that if someone needs this functionality they have to add the frame kernel to their SPICE dependency chart, was this already the case or do we need to do an inventory there?

tmplummer commented:
> Looks good to me. The only downside is that if someone needs this functionality they have to add the frame kernel to their SPICE dependency chart, was this already the case or do we need to do an inventory there?

Agreed, this is a slight downside. I suspect all the processing getting spin-phase already required this kernel, but will need to let the team know about this change on Monday.

tmplummer committed:
Add required spice kernel to instrument spin phase test

tmplummer committed:
Revert back to using fixed lookup table for instrument spin-phase offset
